### PR TITLE
Support fallback to plain http in nerdctl login

### LINKED ISSUE #887: http registry login (`http: server gave HTTP response to HTTPS client`)
I found this question #224 but setting the `--insecure-registry` flag only skips tls verfiy and does not affect the https://github.com/containerd/nerdctl/blob/354d03acb11a1649f0f3aec73218b8d8e2d9e062/pkg/imgutil/dockerconfigresolver/dockerconfigresolver.go#L39 option (anymore?).

I figured out plainHTTP is only accepted on localhost. That's totally fine in terms of security not sending credentials in plain text over the network. Nevertheless I have to do it (even though VPN) ...

Can we provide a new flag for http? Or could we parse the protocol form the host input for example `nerdctl login http://registry.com` ?

### PR DESCRIPTION
**Pull request by Zheaoli**
Fix #887 

Signed-off-by: Zheao.Li <me@manjusaka.me>

### CODE REVIEW

File: pkg/imgutil/imgutil.go [hunk 1]
```diff
@@ -145,7 +146,7 @@ func EnsureImage(ctx context.Context, client *containerd.Client, stdout, stderr
 	img, err := PullImage(ctx, client, stdout, stderr, snapshotter, resolver, ref, ocispecPlatforms, unpack, quiet)
 	if err != nil {
 		// In some circumstance (e.g. people just use 80 port to support pure http), the error will contain message like "dial tcp <port>: connection refused".
-		if !IsErrHTTPResponseToHTTPSClient(err) && !IsErrConnectionRefused(err) {
+		if !common.IsErrHTTPResponseToHTTPSClient(err) && !common.IsErrConnectionRefused(err) {
```
Comment (AkihiroSuda) [inline pkg/imgutil/imgutil.go:149]: `common` -> `errutil`

Comment (Zheaoli) [reply pkg/imgutil/imgutil.go:149]: Got it

File: cmd/nerdctl/login.go [hunk 2]
```diff
@@ -240,14 +241,44 @@ func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthCon
...
 		if err == nil {
 			return identityToken, nil
 		}
+		if insecure && (errutil.IsErrHTTPResponseToHTTPSClient(err) || errutil.IsErrConnectionRefused(err)) {
+			needFallbackToPlainHttp = true
+			break
+		}
 		logrus.WithError(err).WithField("i", i).Error("failed to call tryLoginWithRegHost")
 	}
+	if needFallbackToPlainHttp {
```
Comment (Junnplus) [inline cmd/nerdctl/login.go:257]: I think we can retry directly after 246 lines with plain http.

Comment (Zheaoli) [reply cmd/nerdctl/login.go:257]: Actually, I'm not sure about this.

File: cmd/nerdctl/login.go [hunk 3]
```diff
@@ -240,14 +241,44 @@ func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthCon
 	if len(regHosts) == 0 {
 		return "", fmt.Errorf("got empty []docker.RegistryHost for %q", host)
 	}
+	needFallbackToPlainHttp := false
 	for i, rh := range regHosts {
 		err = tryLoginWithRegHost(ctx, rh)
```
Comment (Junnplus) [inline cmd/nerdctl/login.go:246]:
```suggestion
                err = tryLoginWithRegHost(ctx, rh)
		if err != nil && insecure && (errutil.IsErrHTTPResponseToHTTPSClient(err) || errutil.IsErrConnectionRefused(err)) {
		    rh.Scheme = "http"
		    err = tryLoginWithReghost(ctx, rh)
		}
```

File: cmd/nerdctl/login.go (hunk 2 above)
Comment (Junnplus) [reply cmd/nerdctl/login.go:257]: see https://github.com/containerd/nerdctl/pull/894/files#r827661658

File: cmd/nerdctl/login_linux_test.go [hunk 4]
```diff
@@ -75,3 +76,102 @@ func TestLoginWithSpecificRegHosts(t *testing.T) {
 	}
 
 }
+
+func TestLoginWithPlainHttp(t *testing.T) {
```
Comment (Junnplus) [inline cmd/nerdctl/login_linux_test.go:80]: I'm not sure if this test works, it seems that `registry:2` image can't verify this issue.

I tested it locally through nexus3

File: cmd/nerdctl/login_linux_test.go [hunk 5]
```diff
@@ -75,3 +76,102 @@ func TestLoginWithSpecificRegHosts(t *testing.T) {
...
+	testutil.DockerIncompatible(t)
+	base := testutil.NewBase(t)
+	reg := testregistry.NewAuthWithHttp(base, "admin", "validTestPassword", 5000)
+	testCasesForPort5000 := []struct {
```
Comment (Junnplus) [inline cmd/nerdctl/login_linux_test.go:84]: IMHO, We should merge `testCasesForPort80` with it.

File: cmd/nerdctl/login.go (hunk 3 above)
Comment (Zheaoli) [reply cmd/nerdctl/login.go:246]: I'm not sure about changing the Scheme directly is the best practice

File: cmd/nerdctl/login_linux_test.go (hunk 5 above)
Comment (Zheaoli) [reply cmd/nerdctl/login_linux_test.go:84]: No, if we merge `testCasesForPort80` with `testCasesForPort5000`, we need to start/shut down all images for the test registry for each test case. I'm not sure it's right or not

File: cmd/nerdctl/login_linux_test.go [hunk 6]
```diff
@@ -75,3 +76,102 @@ func TestLoginWithSpecificRegHosts(t *testing.T) {
...
+			regHost := tc.regHost
+			if tc.useRegPort {
+				regHost = fmt.Sprintf("%s:%d", regHost, tc.regPort)
+			}
+			if tc.shouldSuccess {
+				t.Logf("Good password")
+			} else {
+				t.Logf("Bad password")
+			}
+			cmd := base.Cmd("--insecure-registry", "--debug-full", "--hosts-dir", reg.HostsDir, "login", "-u", tc.username, "-p", tc.password, regHost)
```
Comment (Junnplus) [inline cmd/nerdctl/login_linux_test.go:121]: We need to verify that it works without `-insecure-registry` flags.

**Review by Junnplus — CHANGES_REQUESTED** [review]

File: cmd/nerdctl/login_linux_test.go (hunk 5 above)
Comment (Junnplus) [reply cmd/nerdctl/login_linux_test.go:84]: > we need to start/shut down all images for the test registry for each test case

Not necessary, just start them both at the same time.

File: cmd/nerdctl/login.go [hunk 7]
```diff
@@ -240,14 +241,44 @@ func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthCon
...
+		// onFetchRefreshToken is called when tryLoginWithRegHost calls rh.Authorizer.Authorize()
+		onFetchRefreshToken := func(ctx context.Context, s string, req *http.Request) {
+			fetchedRefreshTokens[req.URL.Host] = s
+		}
+		ho.AuthorizerOpts = append(ho.AuthorizerOpts, docker.WithFetchRefreshToken(onFetchRefreshToken))
+		regHosts, err = dockerconfig.ConfigureHosts(ctx, *ho)(host)
+		if err != nil {
+			return "", err
+		}
+		for i, rh := range regHosts {
```
Comment (fahedouch) [inline cmd/nerdctl/login.go:273]: no need to retry all `regHosts` you should fallback to http only in the registry that faced `(errutil.IsErrHTTPResponseToHTTPSClient(err) || errutil.IsErrConnectionRefused(err))`

File: cmd/nerdctl/login_linux_test.go [hunk 8]
```diff
@@ -75,3 +76,100 @@ func TestLoginWithSpecificRegHosts(t *testing.T) {
...
+		{
+			regHost:           "127.0.0.1",
+			regPort:           80,
+			useRegPort:        false,
+			username:          "admin",
+			password:          "invalidTestPassword",
+			shouldSuccess:     false,
+			registry:          reg80,
+			shouldUseInSecure: true,
+		},
```
Comment (fahedouch) [inline cmd/nerdctl/login_linux_test.go:146]: plz add a test `shouldUseInSecure: false` for `registry:  reg80` to check that there is not a regression

Comment (Zheaoli) [reply cmd/nerdctl/login_linux_test.go:146]: OK

File: cmd/nerdctl/login.go (hunk 3 above)
Comment (Junnplus) [reply cmd/nerdctl/login.go:247]: I reserve my opinion. 

When it is not a valid HTTPs request and `insecure-registry` flaged, we can fallback directly to plain HTTP. cc @fahedouch

**Comment by AkihiroSuda** [comment]
Seems reopened as https://github.com/containerd/nerdctl/pull/907 (why?)

**Comment by Zheaoli** [comment]
> Seems reopened as #907 (why?)

I have squashed my commit but not synced in the PR because of the GitHub issue. So I reopened the PR

**Pull request closed without merging**